libpod/runtime.go

```diff
@@ -136,81 +137,244 @@ func (r *Runtime) Shutdown(force bool) error {
+	defer r.lock.Unlock()
+
+	if !r.valid {
+		return nil, fmt.Errorf("runtime has already been shut down")
+	}
+
+	ctr, err := newContainer(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, option := range options {
+		if err := option(ctr); err != nil {
+			return nil, errors.Wrapf(err, "error running container create option")
+		}
+	}
+
+	ctr.valid = true
+
+	if err := r.state.AddContainer(ctr); err != nil {
+		// If we joined a pod, remove ourself from it
+		if ctr.pod != nil {
+			if err2 := ctr.pod.removeContainer(ctr); err2 != nil {
+				return nil, errors.Wrapf(err, "error adding new container to state, container could not be removed from pod %s", ctr.pod.ID())
+			}
+		}
+
+		// TODO: Might be worth making an effort to detect duplicate IDs
+		// We can recover from that by generating a new ID for the
+		// container
+		return nil, errors.Wrapf(err, "error adding new container to state")
+	}
+
+	return ctr, nil
 }
 
 // RemoveContainer removes the given container
 // If force is specified, the container will be stopped first
 // Otherwise, RemoveContainer will return an error if the container is running
-func (r *Runtime) RemoveContainer(c *ctr.Container, force bool) error {
-	return ctr.ErrNotImplemented
+func (r *Runtime) RemoveContainer(c *Container, force bool) error {
+	return errNotImplemented
 }
 
 // GetContainer retrieves a container by its ID
-func (r *Runtime) GetContainer(id string) (*ctr.Container, error) {
-	return nil, ctr.ErrNotImplemented
+func (r *Runtime) GetContainer(id string) (*Container, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	if !r.valid {
+		return nil, fmt.Errorf("runtime has already been shut down")
+	}
+
+	return r.state.GetContainer(id)
+}
+
+// HasContainer checks if a container with the given ID is present
+func (r *Runtime) HasContainer(id string) (bool, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	if !r.valid {
+		return false, fmt.Errorf("runtime has already been shut down")
+	}
+
+	return r.state.HasContainer(id)
 }
 
 // LookupContainer looks up a container by its name or a partial ID
 // If a partial ID is not unique, an error will be returned
-func (r *Runtime) LookupContainer(idOrName string) (*ctr.Container, error) {
-	return nil, ctr.ErrNotImplemented
+func (r *Runtime) LookupContainer(idOrName string) (*Container, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	if !r.valid {
+		return nil, fmt.Errorf("runtime has already been shut down")
+	}
+
+	return r.state.LookupContainer(idOrName)
 }
 
 // GetContainers retrieves all containers from the state
 // Filters can be provided which will determine what containers are included in
 // the output. Multiple filters are handled by ANDing their output, so only
 // containers matching all filters are returned
-func (r *Runtime) GetContainers(filters ...ContainerFilter) ([]*ctr.Container, error) {
-	return nil, ctr.ErrNotImplemented
+func (r *Runtime) GetContainers(filters ...ContainerFilter) ([]*Container, error) {
```

Comment:
Should this just be Containers? Drop the Get?

Reply:
We had GetContainer, so I made this match. I think Containers might be better, though, it's less easy to confuse the two that way.